src/cryptoadvance/specter/util/google_drive.py

```diff
@@ -0,0 +1,137 @@
+import os, io, json
+
+from pathlib import Path
+from dotenv import load_dotenv
+
+from googleapiclient.discovery import build
+from googleapiclient.http import MediaIoBaseUpload, MediaIoBaseDownload
+from google_auth_oauthlib.flow import Flow
+
+from flask import current_app as app, session, request, redirect, url_for
+
+from zipfile import ZipFile
+
+env_path = Path(".") / ".flaskenv"
+load_dotenv(env_path)
+
+# If modifying these scopes, delete the file token.json.
+CLIENT_CONIG = {
+    "web": {
+        "client_id": os.environ.get("GOOGLE_OAUTH_CLIENT_ID"),
+        "project_id": os.environ.get("GOOGLE_OAUTH_PROJECT_ID"),
+        "auth_uri": "https://accounts.google.com/o/oauth2/auth",
+        "token_uri": "https://oauth2.googleapis.com/token",
+        "auth_provider_x509_cert_url": "https://www.googleapis.com/oauth2/v1/certs",
+        "client_secret": os.environ.get("GOOGLE_OAUTH_CLIENT_SECRET"),
+        "redirect_uris": ["http://localhost:25441/"],
+        "javascript_origins": ["http://localhost:25441"],
+    }
+}
+SCOPES = ["https://www.googleapis.com/auth/drive.appdata"]
+REDIRECT_URI = "http://localhost:25441/settings/backup_to_google_drive/callback"
+
+
+def trigger_oauth(current_user):
+    flow = Flow.from_client_config(client_config=CLIENT_CONIG, scopes=SCOPES)
+    flow.redirect_uri = REDIRECT_URI
+    authorization_url, state = flow.authorization_url(
+        # Enable offline access so that you can refresh an access token without
+        # re-prompting the user for permission. Recommended for web server apps.
+        access_type="offline",
+        # Enable incremental authorization. Recommended as a best practice.
+        include_granted_scopes="true",
+    )
+    current_user.set_google_oauth_state(state)
+    return {"redirect_url": authorization_url}
+
+
+def backup(file, current_user):
+    if not "creds" in current_user.google_oauth_data:
+        return trigger_oauth(current_user)
+    try:
+        creds = current_user.google_oauth_data["creds"]
+        service = build("drive", "v3", credentials=creds)
+        file_metadata = {"name": "specter-backup.zip", "parents": ["appDataFolder"]}
+        media = MediaIoBaseUpload(file, mimetype="application/zip", resumable=True)
+        file = (
+            service.files()
+            .create(body=file_metadata, media_body=media, fields="id")
+            .execute()
+        )
+        app.logger.info("Backed up to Google Drive successfully!")
+        response = (
+            service.files()
+            .list(
+                spaces="appDataFolder",
+                fields="nextPageToken, files(id, name)",
+                pageSize=10,
+            )
+            .execute()
+        )
+        return {"success": True}
+    except Exception as e:
```

Comment:
Hmm. I wanted to account for any and every exception that google-drive's apis may throw. Even if I am able to manually try a few possibilities, wouldn't there be some exception that I couldn't find? Shouldn't `except Exception as e:` still be there just in case this happens?

Reply:
The thing is: Do you know how to deal with those exceptions? Maybe yes. But then figure out the exception hirarchy of google. Probably they have a common Exception they are all derived from. That way you would handle Google-Exceptions, not ALL exceptions. That's an important difference as you might cover e.g. an Out-of-memory-Error and assume it's some Google-Exception.

So in general, it's much better to NOT handle an exception and let it propagate then to suppress it and try to move on (if you don't have a rough understanding why that Exception is thrown).

The kind of generic handling is already built in in certain levels of the system:
https://github.com/cryptoadvance/specter-desktop/blob/master/src/cryptoadvance/specter/server_endpoints/controller.py#L49-L87
As you can see, we're handling 3 different errors slightly different. So in your case (and if you would remove your catching-code) if weird Google-exceptions might be thrown, you end up in Line 77 and a generic error-handler-page (and logging in the logs). 
This is the best to figure out what's happening and a proper fix can get delivered based on that information which deals with the details of this issue in a much more informed way,

Rule of thumb: Don't deal with errors you don't know about.

That doesn't mean that you should look out for errors and think about how to deal with them best. But don't assume all Exceptions are Google-Exceptions.

Reply:
That really makes so much sense. Thanks! I'll keep this in mind and make changes.